### PR TITLE
feat: support `stride` in `array_slice`, change indexes to be`1` based

### LINKED ISSUE #8784: Support stride for `array_slice`
### Is your feature request related to a problem or challenge?

Currently, the `array_slice` function only accepts three arguments: the `array`, `from`, and `to`. However, it would be more convenient if it also supported a `stride` parameter.

Expect result
```
array_slice([1, 2, 3, 4], 1, 3, 2) --> [1, 3]
```

### Describe the solution you'd like

_No response_

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Weijun-H**
## Which issue does this PR close?

Closes #8784

## What changes are included in this PR?
support array_slice with stride argument
```
array_slice(make_array(1, 2, 3, 4, 5), 1, 5, 2) --> [1, 3, 5]
```

### CODE REVIEW

File: datafusion/physical-expr/src/array_expressions.rs [hunk 1]
```diff
@@ -701,12 +721,45 @@ where
         if let (Some(from), Some(to)) = (from_index, to_index) {
             if from <= to {
                 assert!(start + to <= end);
-                mutable.extend(
-                    0,
-                    (start + from).to_usize().unwrap(),
-                    (start + to + O::usize_as(1)).to_usize().unwrap(),
-                );
-                offsets.push(offsets[row_index] + (to - from + O::usize_as(1)));
+                if let Some(strde) = strde {
```
Comment (jayzhan211) [inline datafusion/physical-expr/src/array_expressions.rs:724]: is `strde` typo?

File: datafusion/sqllogictest/test_files/array.slt [hunk 2]
```diff
@@ -1214,6 +1214,19 @@ select array_slice(make_array(1, 2, 3, 4, 5), 2, 4), array_slice(make_array('h',
...
+[1, 3, 5] [h, l, o]
+
+query ??
+select array_slice(make_array(1, 2, 3, 4, 5), 1, 5, -1), array_slice(make_array('h', 'e', 'l', 'l', 'o'), 1, 5, -1);
+----
+[] []
+
+query error Execution error: array_slice got invalid stride: 0, it cannot be 0
+select array_slice(make_array(1, 2, 3, 4, 5), 1, 5, 0), array_slice(make_array('h', 'e', 'l', 'l', 'o'), 1, 5, 0);
+
```
Comment (jayzhan211) [inline datafusion/sqllogictest/test_files/array.slt:1229]: `SELECT([1, 2, 3, 4, 5])[4:2:-2];` are also needed. And other examples in duckdb, we should consider them too.

https://duckdb.org/docs/sql/functions/nested#slicing

Comment (Weijun-H) [reply datafusion/sqllogictest/test_files/array.slt:1229]: Yes, I think so. And we could do it in this ticket.

File: datafusion/physical-expr/src/array_expressions.rs [hunk 3]
```diff
@@ -680,6 +689,16 @@ where
         let end = offset_window[1];
         let len = end - start;
 
+        let stride = if let Some(stride) = stride {
```
Comment (comphead) [inline datafusion/physical-expr/src/array_expressions.rs:692]: this construction most likely can be replaced by `.map/.and_then` ?

File: datafusion/physical-expr/src/array_expressions.rs [hunk 4]
```diff
@@ -702,15 +721,75 @@ where
         if let (Some(from), Some(to)) = (from_index, to_index) {
             if from <= to {
                 assert!(start + to <= end);
-                mutable.extend(
-                    0,
-                    (start + from).to_usize().unwrap(),
-                    (start + to + O::usize_as(1)).to_usize().unwrap(),
-                );
-                offsets.push(offsets[row_index] + (to - from + O::usize_as(1)));
+                if let Some(stride) = stride {
```
Comment (comphead) [inline datafusion/physical-expr/src/array_expressions.rs:724]: this check may be earlier?

**Review by comphead — COMMENTED** [review]
Thanks @Weijun-H, just couple of minors, and we need to modify the user documentation?

File: datafusion/expr/src/expr_fn.rs [hunk 5]
```diff
@@ -731,7 +731,7 @@ scalar_expr!(
 scalar_expr!(
     ArraySlice,
     array_slice,
-    array offset length,
+    array offset length stride,
```
Comment (jayzhan211) [inline datafusion/expr/src/expr_fn.rs:734]: I think we should change this to `array begin end step`. step to be consistent with duckdb docs

File: datafusion/physical-expr/src/array_expressions.rs [hunk 6]
```diff
@@ -702,15 +721,75 @@ where
...
+                    mutable.extend(
+                        0,
+                        (start + from).to_usize().unwrap(),
+                        (start + to + O::usize_as(1)).to_usize().unwrap(),
+                    );
+                    offsets.push(offsets[row_index] + (to - from + O::usize_as(1)));
+                }
             } else {
+                let stride = if let Some(stride) = stride {
+                    if !stride.is_negative() {
```
Comment (jayzhan211) [inline datafusion/physical-expr/src/array_expressions.rs:764]: you can use `is_positive()`

**Review by jayzhan211 — APPROVED** [review]
LGTM

**Comment by Weijun-H** [comment]
stalled until #8847 fixed

File: datafusion/physical-expr/src/array_expressions.rs [hunk 7]
```diff
@@ -700,17 +709,67 @@ where
         };
 
         if let (Some(from), Some(to)) = (from_index, to_index) {
+            let stride = stride.map(|s| s.value(row_index));
```
Comment (comphead) [inline datafusion/physical-expr/src/array_expressions.rs:712]: 👍

**Review by comphead — COMMENTED** [review]
Thanks @Weijun-H for the fixes, its much nicer now. Waiting for the CI

**Comment by alamb** [comment]
CI seems broken on this PR

**Comment by alamb** [comment]
> stalled until #8847 fixed

Marking as draft to make it clear this PR is not waiting on review

**Review by alamb — APPROVED** [review]
Looks good to me @Weijun-H  -- thank you

Can you confirm the behavior change to be `0` based indexes rather than `1` based indexes is intentional? I updated the PR title but I wanted to double check

**Comment by Weijun-H** [comment]
> Looks good to me @Weijun-H -- thank you
> 
> Can you confirm the behavior change to be `0` based indexes rather than `1` based indexes is intentional? I updated the PR title but I wanted to double check

`array_slice` is 1-index, not 0-index, which aligns with PostgreSQL and DuckDB. @alamb

**Comment by alamb** [comment]
> > Looks good to me @Weijun-H -- thank you
> > Can you confirm the behavior change to be `0` based indexes rather than `1` based indexes is intentional? I updated the PR title but I wanted to double check
> 
> `array_slice` is 1-index, not 0-index, which aligns with PostgreSQL and DuckDB. @alamb

Fixed

**Pull request merged**